File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/simple/SimpleEntityTest.scala
package com.github.debop4s.data.tests.mapping.simple

import com.github.debop4s.core.io.Serializers
import com.github.debop4s.core.utils.Hashs
import com.github.debop4s.data.jpa.repository.JpaDao
import com.github.debop4s.data.model.LongEntity
import com.github.debop4s.data.tests.AbstractJpaTest
import java.util.Date
import javax.persistence._
import org.hibernate.annotations.CacheConcurrencyStrategy
import org.hibernate.{annotations => ha}
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.transaction.annotation.Transactional


/**
 * Created by debop on 2014. 3. 9.
 */
@Transactional
class SimpleEntityTest extends AbstractJpaTest {

    @Autowired val dao: JpaDao = null

    @Test
    def lifecycle() {
        val entity = new LifecycleEntity()
        entity.name = "이름"
        dao.persist(entity)
        dao.flush()

        dao.em.detach(entity)
        entity.name = "변경된 이름"
        dao.em.merge(entity)
        dao.flush()

        dao.clear()

        val loaded = dao.findOne(classOf[LifecycleEntity], entity.id)
        assert(loaded != null)
        assert(loaded == entity)
        assert(loaded.createdAt != null)
        assert(loaded.updatedAt != null)

        dao.delete(loaded)
        dao.flush()
        dao.clear()

        assert(dao.findOne(classOf[LifecycleEntity], loaded.id) == null)
    }

    @Test
    def transientObject() {
        val transient = new SimpleEntity()
        transient.name = "transient"

        val transient2 = Serializers.copyObject(transient)
        transient2.description = "description"
        assert(transient2 == transient)

        val saved = Serializers.copyObject(transient)
        dao.persist(saved)
        dao.flush()
        dao.clear()

        assert(saved != transient)

        val loaded = dao.findOne(classOf[SimpleEntity], saved.id)
        assert(loaded != null)
        assert(loaded == saved)
        assert(loaded != transient)

        val saved2 = Serializers.copyObject(transient)
        dao.persist(saved2)
        dao.flush()
        dao.clear()

        val loaded2 = dao.findOne(classOf[SimpleEntity], saved2.id)
        assert(loaded2 != null)
        assert(loaded2 == saved2)
        assert(loaded2 != transient)
        assert(loaded2 != loaded)

        dao.delete(loaded2)
        dao.flush()
        dao.clear()

        assert(dao.findOne(classOf[SimpleEntity], saved2.id) == null)


    }
}

@Entity
@org.hibernate.annotations.Cache(region = "simple", usage = CacheConcurrencyStrategy.READ_WRITE)
@ha.DynamicInsert
@ha.DynamicUpdate
class LifecycleEntity extends LongEntity {

    @Index(name = "ix_lifecycleentity_name", columnList = "name")
    var name: String = _

    // NOTE: Higernate @Generated 는 JPA 에서는 작동하지 않는다. JPA에서는 @PrePersist @PreUpdate 를 사용해야 합니다.
    // @ha.Generated(ha.GenerationTime.INSERT)
    @Column(name = "createdAt", updatable = false)
    @Temporal(TemporalType.TIMESTAMP)
    var createdAt: Date = _

    // @ha.Generated(ha.GenerationTime.ALWAYS)
    @Column(name = "updatedAt", insertable = false)
    @Temporal(TemporalType.TIMESTAMP)
    var updatedAt: Date = _

    @PrePersist
    def onPrePersist() {
        if (createdAt == null) createdAt = new Date()
    }

    @PreUpdate
    def onPreUpdate() {
        updatedAt = new Date()
    }

    override def hashCode(): Int = Hashs.compute(name)
}

@Entity
@org.hibernate.annotations.Cache(region = "simple", usage = CacheConcurrencyStrategy.READ_WRITE)
@ha.DynamicInsert
@ha.DynamicUpdate
class SimpleEntity extends LongEntity {

    var name: String = _
    var description: String = _

    override def hashCode(): Int = Hashs.compute(name)
}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/simple/SimpleEntityTest.java
package debop4s.data.orm.jtests.mapping.simple;

import debop4s.core.io.Serializers;
import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.data.orm.s.mapping.simple.SimpleEntityJUnitSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 2. 오후 11:25
 */
@Slf4j
@Transactional
public class SimpleEntityTest extends JpaTestBase {

    @PersistenceContext EntityManager em;

    @Test
    public void lifecycleTest() {

        LifecycleEntity entity = new LifecycleEntity();
        entity.setName("이름");
        em.persist(entity);
        em.flush();

        em.detach(entity);
        entity.setName("갱신");
        em.merge(entity);
        em.flush();
        em.clear();

        LifecycleEntity loaded = em.find(LifecycleEntity.class, entity.getId());
        assertThat(loaded).isNotNull();

        assertThat(loaded.getCreatedAt()).isNotNull();
        assertThat(loaded.getUpdatedAt()).isNotNull();

        em.remove(loaded);
        em.flush();

        assertThat(em.find(LifecycleEntity.class, entity.getId())).isNull();
    }

    @Test
    public void transientObjectTest() {
        SimpleEntity transientObj = new SimpleEntity();
        transientObj.setName("transient");

        SimpleEntity transientObj2 = Serializers.copyObject(transientObj);

        transientObj2.setDescription("desc");
        assertThat(transientObj2).isEqualTo(transientObj);

        SimpleEntity savedObj = Serializers.copyObject(transientObj);
        em.persist(savedObj);
        em.flush();
        em.clear();

        // Id를 발급받은 Persistent Object 와 Transient Object 와의 비교. hashCode에서
        assertThat(savedObj).isNotEqualTo(transientObj);

        SimpleEntity loaded = em.find(SimpleEntity.class, savedObj.getId());

        assertThat(loaded).isNotNull();
        // Persistent Object 간의 비교
        assertThat(loaded).isEqualTo(savedObj);

        // Persistent Object 와 Transient Object 간의 비교
        assertThat(loaded).isNotEqualTo(transientObj);

        SimpleEntity savedObj2 = Serializers.copyObject(transientObj);
        em.persist(savedObj2);
        em.flush();
        em.clear();

        SimpleEntity loaded2 = em.find(SimpleEntity.class, savedObj2.getId());
        assertThat(loaded2).isNotNull();
        assertThat(loaded2).isEqualTo(savedObj2);
        assertThat(loaded2).isNotEqualTo(transientObj);
        assertThat(loaded2).isNotEqualTo(loaded);

        em.remove(loaded2);
        em.flush();

        assertThat(em.find(SimpleEntity.class, savedObj2.getId())).isNull();
    }
}
